File: ring-android/app/src/main/java/cx/ring/fragments/QRCodeFragment.java
/*
 * Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 * Authors:    AmirHossein Naghshzan <amirhossein.naghshzan@savoirfairelinux.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package cx.ring.fragments;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import cx.ring.R;
import cx.ring.databinding.FragQrcodeBinding;
import cx.ring.share.ScanFragment;
import cx.ring.share.ShareFragment;
import cx.ring.utils.DeviceUtils;

public class QRCodeFragment extends BottomSheetDialogFragment {

    public static final String TAG = QRCodeFragment.class.getSimpleName();
    public static final String ARG_START_PAGE_INDEX = "start_page";

    public static final int INDEX_CODE = 0;
    public static final int INDEX_SCAN = 1;

    public static QRCodeFragment newInstance(int startPage) {
        QRCodeFragment fragment = new QRCodeFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_START_PAGE_INDEX, startPage);
        fragment.setArguments(args);
        return fragment;
    }

    private FragQrcodeBinding mBinding = null;
    private int mStartPageIndex;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        Bundle args = getArguments();
        mStartPageIndex = args.getInt(ARG_START_PAGE_INDEX, 0);

        mBinding = FragQrcodeBinding.inflate(inflater, container, false);
        mBinding.viewPager.setAdapter(new SectionsPagerAdapter(getContext(), getChildFragmentManager()));
        mBinding.tabs.setupWithViewPager(mBinding.viewPager);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (mStartPageIndex != 0) {
            mBinding.tabs.getTabAt(mStartPageIndex).select();
        }
    }

    @Override
    public void onDestroyView() {
        mBinding = null;
        super.onDestroyView();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.setOnShowListener(dialogINterface -> {
            if (DeviceUtils.isTablet(requireContext())) {
                dialog.getWindow().setLayout(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.MATCH_PARENT);
            }
        });
        return dialog;
    }

    static class SectionsPagerAdapter extends FragmentPagerAdapter {
        @StringRes
        private final int[] TAB_TITLES = new int[]{R.string.tab_code, R.string.tab_scan};
        private final Context mContext;

        SectionsPagerAdapter(Context context, FragmentManager fm) {
            super(fm);
            mContext = context;
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    return new ShareFragment();
                case 1:
                    return new ScanFragment();
                default:
                    return null;
            }
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return mContext.getResources().getString(TAB_TITLES[position]);
        }

        @Override
        public int getCount() {
            return TAB_TITLES.length;
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        addGlobalLayoutListener(getView());
    }

    private void addGlobalLayoutListener(final View view) {
        view.addOnLayoutChangeListener(new View.OnLayoutChangeListener() {
            @Override
            public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                setPeekHeight(v.getMeasuredHeight());
                v.removeOnLayoutChangeListener(this);
            }
        });
    }

    public void setPeekHeight(int peekHeight) {
        BottomSheetBehavior<?> behavior = getBottomSheetBehaviour();
        if (behavior == null) {
            return;
        }

        behavior.setPeekHeight(peekHeight);
    }

    private BottomSheetBehavior<?> getBottomSheetBehaviour() {
        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) ((View) getView().getParent()).getLayoutParams();
        CoordinatorLayout.Behavior<?> behavior = layoutParams.getBehavior();
        if (behavior instanceof BottomSheetBehavior) {
            return (BottomSheetBehavior<?>) behavior;
        }

        return null;
    }

}

File: jami-android/app/src/main/java/cx/ring/fragments/QRCodeFragment.kt
/*
 *  Copyright (C) 2004-2025 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.fragments

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import cx.ring.R
import cx.ring.databinding.FragQrcodeBinding
import cx.ring.share.ScanFragment
import cx.ring.share.ShareFragment
import net.jami.model.Uri

class QRCodeFragment : BottomSheetDialogFragment() {
    private var mBinding: FragQrcodeBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)

        val mode = arguments?.getInt(ARG_MODE) ?: 0
        val startMode = arguments?.getInt(ARG_START_MODE) ?: 0
        val contactUri = Uri.fromString(arguments?.getString(ARG_CONTACT_URI) ?: "")

        return FragQrcodeBinding.inflate(inflater, container, false).apply {
            when (mode) { // Check on mode
                MODE_SCAN, MODE_SHARE -> {}
                MODE_SCAN or MODE_SHARE ->
                    if (startMode == 0) throw IllegalArgumentException("Start mode not set")

                0 -> throw IllegalArgumentException("Mode not set")
                else -> throw IllegalArgumentException("Unknown mode: $mode")
            }

            SectionsPagerAdapter(this@QRCodeFragment).apply {
                // Add fragments based on mode
                if (mode and MODE_SHARE != 0) {
                    ShareFragment.newInstance(contactUri).apply {
                        addFragment(this, getTabTitle(INDEX_SHARE))
                    }
                }
                if (mode and MODE_SCAN != 0) addFragment(ScanFragment(), getTabTitle(INDEX_SCAN))
                viewPager.adapter = this

                // If multi-mode, setup the tabs. If not, hide them.
                if (mode and MODE_SHARE != 0 && mode and MODE_SCAN != 0) {
                    viewPager.currentItem = when (startMode) {
                        MODE_SHARE -> INDEX_SHARE
                        MODE_SCAN -> INDEX_SCAN
                        else -> INDEX_SCAN
                    }
                    val tabs: TabLayout = tabs
                    TabLayoutMediator(tabs, viewPager) { tab, position ->
                        tab.text = getTabTitle(position)
                    }.attach()
                } else tabs.visibility = View.GONE
            }
            mBinding = this
        }.root
    }

    override fun onDestroyView() {
        mBinding = null
        super.onDestroyView()
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        (dialog as BottomSheetDialog).behavior.apply {
            state = BottomSheetBehavior.STATE_EXPANDED
            skipCollapsed = true
        }
        return dialog
    }

    override fun onResume() {
        super.onResume()
        addGlobalLayoutListener(requireView())
    }

    private fun addGlobalLayoutListener(view: View) {
        view.addOnLayoutChangeListener(object : View.OnLayoutChangeListener {
            override fun onLayoutChange(
                v: View, left: Int, top: Int, right: Int, bottom: Int,
                oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int,
            ) {
                setPeekHeight(v.measuredHeight)
                v.removeOnLayoutChangeListener(this)
            }
        })
    }

    private fun setPeekHeight(peekHeight: Int) {
        bottomSheetBehaviour?.peekHeight = peekHeight
    }

    private val bottomSheetBehaviour: BottomSheetBehavior<*>?
        get() {
            val layoutParams =
                (requireView().parent as View).layoutParams as CoordinatorLayout.LayoutParams
            val behavior = layoutParams.behavior
            return if (behavior is BottomSheetBehavior<*>) {
                behavior
            } else null
        }

    internal class SectionsPagerAdapter(val fragment:  Fragment) :
        FragmentStateAdapter(fragment) {
        private val mFragmentList: MutableList<Fragment> = ArrayList()
        private val mFragmentTitleList: MutableList<String> = ArrayList()

        fun getTabTitle(position: Int): String =
            fragment.resources.getString(TAB_TITLES[position])

        fun addFragment(fragment: Fragment, title: String) {
            mFragmentList.add(fragment)
            mFragmentTitleList.add(title)
        }

        override fun getItemCount() = mFragmentList.size

        override fun createFragment(position: Int) = mFragmentList[position]

        companion object {
            @StringRes
            private val TAB_TITLES = intArrayOf(R.string.tab_code, R.string.tab_scan)
        }
    }

    companion object {

        val TAG = QRCodeFragment::class.simpleName!!
        const val ARG_MODE = "mode"
        const val ARG_START_MODE = "start_mode"
        const val ARG_CONTACT_URI = "contact_uri"
        const val INDEX_SHARE = 0
        const val INDEX_SCAN = 1

        const val MODE_SCAN = 1
        const val MODE_SHARE = 2

        @IntDef(flag = true, value = [MODE_SCAN, MODE_SHARE])
        @Retention(AnnotationRetention.SOURCE)
        annotation class QRCodeMode

        /**
         * Create a new instance of QRCodeFragment.
         * @param mode The mode of the fragment.
         * @param startPage The start page of the fragment. Can be null if mode is not multi-mode.
         * @param contactUri The contact URI to share. Can be null if mode is not share.
         */
        fun newInstance(
            @QRCodeMode mode: Int,
            @QRCodeMode startPage: Int? = null,
            contactUri: Uri? = null,
        ): QRCodeFragment {
            val fragment = QRCodeFragment()
            val args = Bundle()
            args.putInt(ARG_MODE, mode)
            if (startPage != null) args.putInt(ARG_START_MODE, startPage)
            if (contactUri != null) args.putString(ARG_CONTACT_URI, contactUri.uri)
            fragment.arguments = args
            return fragment
        }
    }
}